batch/src/main/java/org/openmrs/analytics/FhirSearchUtil.java

```diff
@@ -0,0 +1,113 @@
+      String jsessionId, boolean parse, FhirContext fhirContext) {
+    try {
+      // TODO switch to HAPI FHIR client if all required features are supported.
+      URIBuilder uriBuilder = new URIBuilder(baseSearchUrl);
+      HttpUriRequest request = RequestBuilder
+          .get()
+          .setUri(uriBuilder.build())
+          .addHeader("Content-Type", "application/fhir+json")
+          .addHeader("Accept-Charset", "utf-8")
+          .addHeader("Accept", "application/fhir+json")
+          .addHeader("Cookie", "JSESSIONID=" + jsessionId)
+          .addParameter("_count", String.valueOf(count))
+          .addParameter("_getpagesoffset", String.valueOf(first))
+          .addParameter("_summary", "data")
+          .build();
+      log.info("Fetching " + request.getURI());
+      String bundleStr = fhirStoreUtil.executeRequest(request);
+      if (parse) {
+        IParser parser = fhirContext.newJsonParser();
+        Bundle bundle = parser.parseResource(Bundle.class, bundleStr);
+        log.info(String.format("Fetched %d resources out of a total of %d",
+            bundle.getEntry().size(), bundle.getTotal()));
+        return bundle;
+      } else {
+        return null;
+      }
+    } catch (URISyntaxException e) {
+      log.error("Malformed FHIR url: " + baseSearchUrl + " Exception: " + e);
+    }
+    return null;
+  }
+
+  String findBaseSearchUrl(Bundle searchBundle) {
+    String searchLink = null;
+    for (BundleLinkComponent link : searchBundle.getLink()) {
+      // TODO: Find and use FHIR constants.
+      if (link.getRelation().equals("self") && searchLink == null) {
+        searchLink = link.getUrl();
+      }
+      if (link.getRelation().equals("next")) {
+        searchLink = link.getUrl();
+        break;
+      }
+    }
+    if (searchLink == null) {
+      throw new IllegalArgumentException(
+          String.format("No proper link information in bundle %s", searchBundle));
+    }
+    try {
+      URI searchUri = new URI(searchLink);
+      NameValuePair pagesParam = null;
+      for (NameValuePair pair : URLEncodedUtils.parse(searchUri, StandardCharsets.UTF_8)) {
+        if (pair.getName().equals("_getpages")) {
+          pagesParam = pair;
+        }
+      }
+      if (pagesParam == null) {
+        throw new IllegalArgumentException(
+            String.format("No _getpages parameter found in search link %s", searchLink));
+      }
+      return new URI(searchUri.getScheme(),
+          searchUri.getAuthority(),
+          searchUri.getPath(),
+          pagesParam.toString(), // Only keep the _getpages parameter.
+          searchUri.getFragment()).toString();
+    } catch (URISyntaxException e) {
+      throw new IllegalArgumentException(
+          String.format("Malformed link information with error %s in bundle %s", e.getMessage(),
+              searchBundle));
+    }
+  }
+
+  public void uploadBundleToCloud(Bundle bundle) {
```

Comment:
If we want to generalize the sink, we could implement this method for different types of sink targets, since most will involve somehow sending over a Bundle

Reply:
Exactly.